File: programmers/Level 2/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;

void dfs(vector<int> &v, int target, int sum, int count){
    if(count != v.size()){
        dfs(v, target, sum + v[count], count + 1);
        dfs(v, target, sum - v[count], count + 1);
    }
    else if(target == sum)
        answer++;
}

int solution(vector<int> numbers, int target) {
    dfs(numbers, target, 0, 0);
    return answer;
}

File: programmers/Level 2/타겟 넘버.py
answer = 0

def dfs(numbers, idx, result):
    if idx == len(numbers): # 탐색을 완료했다면,
        if result == key: # 타겟 넘버라면,
            global answer
            answer += 1
        return

    number = numbers[idx]
    dfs(numbers, idx + 1, result + number)
    dfs(numbers, idx + 1, result - number)

def solution(numbers, target):
    global key, answer
    key = target

    dfs(numbers, 0, 0)

    return answer
